### ISSUE TITLE
External Libraries need upgrade

### ISSUE THREAD

**Issue by flyfire100** [python-pillow/Pillow#4234]
when  i  use pillow, the external so is out of date, there are some CVE in this so

**Comment by radarhere** [comment]
Which library/libraries are you referring to?

**Comment by flyfire100** [comment]
zlib | 1.2.3 | CVE-2016-9843
zlib | 1.2.3 | CVE-2016-9842
zlib | 1.2.3 | CVE-2016-9841
zlib | 1.2.3 | CVE-2016-9840

bzip2 | 1.0.6 | CVE-2016-3189
bzip2 | 1.0.6 | CVE-2019-12900


libjpeg | 9c | CVE-2018-11813
libjpeg | 9c | CVE-2018-10126


libpng | 1.6.37 | CVE-2018-14048
libpng | 1.6.37 | CVE-2019-6129


libtiff | 4.0.10 | CVE-2018-17000
libtiff | 4.0.10 | CVE-2019-6128
libtiff | 4.0.10 | CVE-2019-17546
libtiff | 4.0.10 | CVE-2019-14973
libtiff | 4.0.10 | CVE-2018-19210
libtiff | 4.0.10 | CVE-2019-7663

**Comment by radarhere** [comment]
How are you installing Pillow? What operating system are you using?

If you are installing Pillow on macOS or Linux from a binary wheel, then yes, we are responsible for these dependencies. If you are installing another way, then it's not within our control - I can't say whose control it is within without more information about your installation method, but I would suspect it lies with your package manager.

- zlib - CVE-2016-9843, CVE-2016-9842, CVE-2016-9841 and CVE-2016-9840 refer to 1.2.8. zlib 1.2.9 or greater has been in Pillow since Pillow 4.1.0.
- bzip2 - CVE-2016-3189 and CVE-2019-12900 refer to 1.0.6. On macOS, the built-in bzip2 is used, so I presume that's 1.0.6. On Linux, we are using 1.0.6. I have started the process of upgrading bzip2 with https://github.com/matthew-brett/multibuild/pull/282
- libjpeg - CVE-2018-11813 refers to 9c. libjpeg 9c is the latest release of libjpeg, so that lies with libjpeg to release a new version with a fix.
- libpng - CVE-2019-6129 refers to 1.6.36, and CVE-2018-14048 refers to 1.6.34. libpng 1.6.37 was part of the Pillow wheels as of Pillow 6.1.0.
- libtiff - CVE-2018-17000, CVE-2018-19210 and CVE-2018-10126 refer to 4.0.9, CVE-2019-6128, CVE-2019-17546 CVE-2019-14973 and CVE-2019-7663 refer to 4.0.10. libtiff 4.0.10 has been part of Pillow since 5.4.0, libtiff 4.1.0 was made available after our latest release, so it will be part of Pillow 7.0 on January 1.

So in summary, if you are using binary wheels on macOS or Linux from the latest version of Pillow
- zlib and libpng should not be a problem
- bzip2 is out-of-date, and I have started to address that
- libjpeg just doesn't have a stable release with fixes
- libtiff will have some of the vulnerabilities that you mentioned. The rest should be fixed with the release of Pillow 7.0 on January 1. This is just due to the nature of our release cycle.

If you are this security conscious, you do also have the option of compiling Pillow from source, meaning that you can install the latest dependencies yourself.

**Comment by flyfire100** [comment]
i install pillow with "pip install pillow"
Our products are based on centos7.2
Thanks for your reply!
